consensus/src/chained_bft/block_storage/pending_votes.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::chained_bft::common::Round;
+use crate::chained_bft::consensus_types::timeout_certificate::TimeoutCertificate;
+use crate::chained_bft::{
+    block_storage::VoteReceptionResult,
+    common::Author,
+    consensus_types::{quorum_cert::QuorumCert, vote_data::VoteData, vote_msg::VoteMsg},
+};
+use crypto::{hash::CryptoHash, HashValue};
+use std::{collections::HashMap, sync::Arc};
+use types::crypto_proxies::LedgerInfoWithSignatures;
+
+struct LastVoteInfo {
+    li_digest: HashValue,
+    round: Round,
+}
+
+/// Last pending votes of the authors. Should be cleared upon reconfiguration.
+pub struct PendingVotes {
+    /// `li_digest_to_votes` might keep multiple LedgerInfos per proposed block in order
+    /// to tolerate non-determinism in execution: given a proposal, a QuorumCertificate is going
+    /// to be collected only for all the votes that carry identical LedgerInfo.
+    /// LedgerInfo digest covers the potential commit ids, as well as the vote information
+    /// (including the 3-chain of a voted proposal).
+    /// Thus, the structure of `li_digest_to_votes` is as follows:
+    /// HashMap<ledger_info_digest, LedgerInfoWithSignatures>
+    li_digest_to_votes: HashMap<HashValue, LedgerInfoWithSignatures>,
+    /// Tracks all the signatures of the votes for the given round. In case we succeed to
+    /// aggregate 2f+1 signatures for the same round a TimeoutCertificate is formed.
+    /// Note that QuorumCert has higher priority than TimeoutCertificate (in case 2f+1 votes are
+    /// gathered for the same ledger info we are going to generate QuorumCert and not the
+    /// TimeoutCertificate).
+    round_to_tc: HashMap<Round, TimeoutCertificate>,
+    /// Map of Author to last vote info. Any pending vote from Author is cleaned up
+    /// whenever new vote is added by same Author
+    author_to_last_voted_info: HashMap<Author, LastVoteInfo>,
+}
+
+impl PendingVotes {
+    pub fn new() -> Self {
+        PendingVotes {
+            li_digest_to_votes: HashMap::new(),
+            round_to_tc: HashMap::new(),
+            author_to_last_voted_info: HashMap::new(),
+        }
+    }
+
+    pub fn insert_vote(
+        &mut self,
+        vote_msg: &VoteMsg,
+        min_votes_for_qc: usize,
+    ) -> VoteReceptionResult {
```

Comment:
not related to this pr, but i don't think `VoteReceptionResult` is more useful than a Result<TC/QC>

Reply:
Yea, I don't think we want to modify it in this PR. Don't have strong opinions. We can change it later.